src/capwap_loc_provider.erl

```diff
@@ -0,0 +1,151 @@
+    when is_integer(RefreshTime) andalso RefreshTime>0 ->
+	validate_providers(Providers),
+	ok.
+
+validate_providers(Providers) when is_list(Providers) ->
+    lists:map(fun validate_provider/1, Providers),
+        ok.
+
+validate_provider({Mod, Config}) ->
+    Mod:validate_provider(Config).
+
+init(Config = #{refresh := RefreshTime, providers := Providers}) ->
+    validate(Config),
+    ?LOG(debug, "table ref: ~p", [ets:whereis(?CACHE_LOC)]),
+    TabOpts = [set, protected, named_table, {read_concurrency, true}],
+    %% Not needed since it will return the table name
+    LocTab = ets:new(?CACHE_LOC, TabOpts),
+    ?LOG(debug, "cache started: ~p", [LocTab]),
+    ?LOG(debug, "table ref: ~p", [ets:whereis(?CACHE_LOC)]),
+    %% TODO Handle exceptions in chain?
+    {ok, #loc_cache{
+	loc_tab = LocTab, refresh = RefreshTime, providers = Providers, provider_chain = chain(Providers)}}.
+
+handle_call(flush_loc_cache, _, State = #loc_cache{loc_tab = LocTab}) ->
+    ?LOG(debug, "Flush location cache, tableref: ~p", [ets:whereis(LocTab)]),
+    ets:delete_all_objects(LocTab),
+    {reply,ok,State};
+handle_call({get_loc, UseCache = true, Name}, _, State = #loc_cache{
+    loc_tab = LocTab, refresh = RefreshTime, provider_chain = Chain}) ->
+	?LOG(debug, "get location for device name (with cache): ~p", [Name]),
+    Result = case ets:lookup(LocTab, Name) of
+	[{Name, Loc = {location, _, _}}] -> Loc;
+	[] -> refresh_loc(Name, Chain, LocTab, RefreshTime)
+    end,
+    {reply, Result, State};
+handle_call({get_loc, UseCache = false, Name}, _, State = #loc_cache{
+	loc_tab = LocTab, refresh = RefreshTime, provider_chain = Chain}) ->
+    ?LOG(debug, "get location for device name (without cache): ~p", [Name]),
+    Result = Chain(Name),
+    {reply, Result, State};
+handle_call({load_config, #{refresh := RefreshTime, providers := Providers}}, _, OldState) ->
+    NewState = OldState#loc_cache{refresh = RefreshTime, providers = Providers, provider_chain = chain(Providers)},
+    ?LOG(debug, "reloading config: from ~p to ~p", [OldState#loc_cache{provider_chain = undefined}, NewState]),
+    {reply,OldState,NewState}.
+
+handle_cast(_, State) ->
+    ?LOG(error, "Unexpected cast"),
+    {noreply,State}.
+
+handle_info({evict, Name}, State = #loc_cache{loc_tab = LocTab}) ->
+    ?LOG(debug, "refresh location expired for device name: ~p", [Name]),
+    ets:delete(LocTab, Name),
+    {noreply,State}.
+
+%% Should cause be logged? Or maybe passed on to the next item?
+prov_fun(ModFun, Next) ->
+    fun(Name) -> case ModFun(Name) of {error, _Cause} -> Next(Name); L = {location, _, _} -> L end end.
+error_fun(_) ->
+    {error, no_more_providers}.
+
+chain(Providers) ->
+    lists:foldl(fun({Mod, Config}, Next) -> prov_fun(Mod:config_fun(Config), Next) end, fun error_fun/1, lists:reverse(Providers)).
+
+terminate(Reason, _) ->
+    ?LOG(debug, "terminating: ~p", [Reason]),
+    ok.
+
+%% Internal functions
```

Comment:
the `refresh_loc/4` is not only one internal function - please or remove this comment or collect all others internal functions below of this comment

Reply:
moved all internal functions downwards

Reply:
cool, thanks - what about `chain/1` `error_fun/1` `prov_fun/2`?

Reply:
True, sorry. Done.